src/coreclr/nativeaot/System.Private.CoreLib/src/System/Collections/Generic/Comparer.CoreRT.cs

```diff
@@ -14,13 +14,22 @@ public abstract partial class Comparer<T> : IComparer, IComparer<T>
     {
         private static Comparer<T> s_default;
 
+        // The AOT compiler can flip this to false under certain circumstances.
+        private static bool SupportsGenericIComparableInterfaces => true;
+
         [Intrinsic]
         private static Comparer<T> Create()
         {
             // The compiler will overwrite the Create method with optimized
             // instantiation-specific implementation.
             // This body serves as a fallback when instantiation-specific implementation is unavailable.
-            Interlocked.CompareExchange(ref s_default, Unsafe.As<Comparer<T>>(ComparerHelpers.GetComparer(typeof(T).TypeHandle)), null);
+            // If that happens, the compiler ensures we generate data structures to make the fallback work
+            // when this method is compiled.
+            Interlocked.CompareExchange(ref s_default,
+                SupportsGenericIComparableInterfaces
+                ? Unsafe.As<Comparer<T>>(ComparerHelpers.GetComparer(typeof(T).TypeHandle))
+                : new ObjectComparer<T>(),
```

Comment:
Could you please double check that there are no unreachable code or structures left around when `SupportsGenericIComparableInterfaces` is true? 

Maybe we should also do the trimming when `SupportsGenericIComparableInterfaces` is true.

Reply:
Good call. We were generating a bunch of useless `ObjectEqualityComparer<Foo>` data structures because the scanner is computing generic dictionaries and scanner isn't inlining the `SupportsGenericIComparableInterfaces`. Fixed.